runtime/signals.c

```diff
@@ -141,16 +163,20 @@ CAMLexport void (*caml_leave_blocking_section_hook)(void) =
 
 CAMLexport void caml_enter_blocking_section(void)
 {
-  while (1){
+  caml_domain_state * domain = Caml_state;
+  while (1) {
     /* Process all pending signals now */
-    caml_raise_if_exception(caml_process_pending_signals_exn());
+    if (Caml_check_gc_interrupt(domain) || domain->action_pending) {
+      /* First reset young_limit, and set action_pending in case there
+         are further async callbacks pending beyond OCaml signal
+         handlers. */
+      caml_handle_gc_interrupt();
+      caml_raise_if_exception(caml_process_pending_signals_exn());
+    }
     caml_enter_blocking_section_hook ();
-    /* Check again for pending signals.
-       If none, done; otherwise, try again */
-    // FIXME: does this become very slow if a signal is recorded but
-    // is masked for everybody in capacity of running signals at this
-    // point?
-    if (!caml_check_pending_signals()) break;
+    /* Check again if a signal arrived in the meanwhile. If none,
+       done; otherwise, try again. */
+    if (atomic_load_relaxed(&domain->young_limit) != UINTNAT_MAX) break;
```

Comment:
I am not very fond of spreading more the (relatively) internal knowledge that `young_limit == UINTNATMAX` means interrupt.
In some places it can't be avoided (runtime assembly changes), in this case maybe a macro would fit better.
Especially since the comment suggests that `young_limit == UINTNAT_MAX` means a signal arrived: it could be in fact a minor/major cycle?
There's also a slight distinction between `Caml_check_gc_interrupt` and checking for `UINTNAT_MAX`: the former has a broader meaning (could be that we reached the end of the minor segment, or we hit an interrupt).
I do think this code segment is fine, but the presence of two form to check for interrupts looks a bit strange to me, maybe I am missing a subtlety.

Reply:
The reason is simple: one can call `Caml_check_gc_interrupt` only when holding the domain lock, since it wants to read `young_ptr`. `UINTNAT_MAX` just appeared to be a convenient value to compare to, there is no tricky mechanism hidden there.